### PR TITLE
Added ThunderPack board (STM32L0xx)

### PR DESCRIPTION
**Pull request by jgillick**
Also:
- Add dfu-util as a upload option to platform.txt.
- https://github.com/jgillick/ThunderPack

This PR implements the following features:

* [ ] New board: ThunderPack (STM32L0xx)
* [ ] ~Add dfu-util as a upload tool to platform.txt.~

~Adding `dfu-util` to `platform.txt` as an upload tool means that users don't need to install the STM32CubeProgrammer. (fewer outside dependencies) The great thing is that Arduino_Core_STM32 already has `dfu-util` built-in, so this is purely a configuration update.~

**Example sketches** 
https://github.com/jgillick/ThunderPack/tree/master/examples/Arduino

### CODE REVIEW

File: boards.txt [hunk 1]
```diff
@@ -1991,3 +1991,25 @@ Midatronics.menu.rtlib.nanofps=Newlib Nano + Float Printf/Scanf
 Midatronics.menu.rtlib.nanofps.build.flags.ldspecs=--specs=nano.specs -u _printf_float -u _scanf_float
 Midatronics.menu.rtlib.full=Newlib Standard
 Midatronics.menu.rtlib.full.build.flags.ldspecs=
+
+###############################
+# ThunderPack
+ThunderPack.name=ThunderPack
```
Comment (fpistm) [inline boards.txt]: Please create a new `Generic L0 series` menu with ThunderPack as a submenu like this further L0 based board addition could be add here.


```suggestion
GenL0.name=Generic STM32L0 series
...
```
then 
```
GenL0.menu.pnum.THUNDER_PACK=ThunderPack
...
```

File: boards.txt [hunk 2]
```diff
@@ -1991,3 +1991,25 @@ Midatronics.menu.rtlib.nanofps=Newlib Nano + Float Printf/Scanf
...
+ThunderPack.build.mcu=cortex-m0plus
+ThunderPack.build.series=STM32L0xx
+ThunderPack.build.cmsis_lib_gcc=arm_cortexM0l_math
+ThunderPack.build.vid=0x0483
+ThunderPack.build.extra_flags={build.usb_flags} -DUSBD_USE_CDC -D{build.product_line} {build.enable_usb} {build.xSerial} -D__CORTEX_SC=0
+
+ThunderPack.upload.maximum_size=196608
+ThunderPack.upload.maximum_data_size=20480
+
+ThunderPack.menu.upload_method.dfuUtil=DFU Util
```
Comment (fpistm) [inline boards.txt]: Please do a separate PR for this.
Note that `linux64` directory has been removed as Arduino IDE does not manage it at runtime.
So you will have to provide a script to properly call the right binary depending of the arch.

File: variants/THUNDER_PACK/variant.h [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+#define PB8  13
+#define PB4  14 // User button
+
+#define PA11 17 // USB_D-
+#define PA12 18 // USB_D+
+#define PA13 19 // SWD
+#define PA14 20 // SWD
+
+// This must be a literal
+#define NUM_DIGITAL_PINS        19
```
Comment (fpistm) [inline variants/THUNDER_PACK/variant.h]:
```suggestion
#define NUM_DIGITAL_PINS        21
```

File: boards.txt (hunk 2 above)
Comment (jgillick) [reply boards.txt]: ✔️

File: boards.txt (hunk 1 above)
Comment (jgillick) [reply boards.txt]: ✔️

File: variants/THUNDER_PACK/variant.h (hunk 3 above)
Comment (jgillick) [reply variants/THUNDER_PACK/variant.h]: ✔️

File: boards.txt (hunk 2 above)
Comment (jgillick) [reply boards.txt]: That's good to know. Is there a good model of this kind of script I should follow? I assume this would go in the `Arduino_Tools` repository?

Comment (fpistm) [reply boards.txt]: Yes, it has been updated on the master and available on next release

**Comment by fpistm** [comment]
Thanks for the update.
Could you fix the astyle issue:
https://travis-ci.com/stm32duino/Arduino_Core_STM32/jobs/247163911

And add all GenL0 options, simply duplicate GenF0 ones in the boards.txt and rename them.
Thanks

File: boards.txt (hunk 2 above)
Comment (jgillick) [reply boards.txt]: Great. Per your recommendation: https://github.com/stm32duino/Arduino_Tools/pull/46

File: variants/THUNDER_PACK/variant.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+#endif
+
+/**
+  * @brief  System Clock Configuration
+  * @param  None
+  * @retval None
+  */
+WEAK void SystemClock_Config(void)
+{
+  RCC_OscInitTypeDef RCC_OscInitStruct = {0};
```
Comment (fpistm) [inline variants/THUNDER_PACK/variant.cpp]: To avoid warnings:
```suggestion
  RCC_OscInitTypeDef RCC_OscInitStruct = {};
```

File: variants/THUNDER_PACK/variant.cpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  RCC_ClkInitTypeDef RCC_ClkInitStruct = {0};
```
Comment (fpistm) [inline variants/THUNDER_PACK/variant.cpp]: To avoid warnings:
```suggestion
  RCC_ClkInitTypeDef RCC_ClkInitStruct = {};
```

File: variants/THUNDER_PACK/variant.cpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  RCC_PeriphCLKInitTypeDef PeriphClkInit = {0};
```
Comment (fpistm) [inline variants/THUNDER_PACK/variant.cpp]: To avoid warnings:
```suggestion
  RCC_PeriphCLKInitTypeDef PeriphClkInit = {};
```

File: variants/THUNDER_PACK/variant.h [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+// This must be a literal
+#define NUM_DIGITAL_PINS        21
+
+// This must be a literal with a value less than or equal to MAX_ANALOG_INPUTS
+#define NUM_ANALOG_INPUTS       4
+#define NUM_ANALOG_FIRST        4
+
+// On-board LED pin number
+#define LED_BUILTIN             0
```
Comment (fpistm) [inline variants/THUNDER_PACK/variant.h]: I advise to use PYn notation to be more precise and inline with notation on you silkscreen.
Moreover it will avoid misalignment if you change the PYn value (pin ordering)

File: variants/THUNDER_PACK/variant.h [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+#define LED_GREEN               LED_BUILTIN
+#define LED_1                   0
```
Comment (fpistm) [inline variants/THUNDER_PACK/variant.h]: dito

File: variants/THUNDER_PACK/variant.h [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+#define LED_2                   1
```
Comment (fpistm) [inline variants/THUNDER_PACK/variant.h]: dito

File: variants/THUNDER_PACK/variant.h [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+#define LED_3                   2
```
Comment (fpistm) [inline variants/THUNDER_PACK/variant.h]: dito

File: variants/THUNDER_PACK/variant.h [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+#define LED_4                   3
```
Comment (fpistm) [inline variants/THUNDER_PACK/variant.h]: dito

File: variants/THUNDER_PACK/variant.h [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+// Use TIM6/TIM7 when possible as servo and tone don't need GPIO output pin
+#define TIMER_TONE              TIM6
+#define TIMER_SERVO             TIM7
+
+// UART Definitions
+#define SERIAL_UART_INSTANCE    1 //ex: 2 for Serial2 (USART2)
+
+// Default pin used for 'Serial' instance (ex: ST-Link)
+// Mandatory for Firmata
+#define PIN_SERIAL_RX           12
```
Comment (fpistm) [inline variants/THUNDER_PACK/variant.h]: dito

File: variants/THUNDER_PACK/variant.h [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+#define PIN_SERIAL_TX           11
```
Comment (fpistm) [inline variants/THUNDER_PACK/variant.h]: dito

File: boards.txt [hunk 14]
```diff
@@ -1009,6 +1009,46 @@ GenF4.menu.upload_method.hidMethod.upload.tool=hid_upload
...
 GenF4.menu.upload_method.hidMethod.build.bootloader_flags=-DBL_HID -DVECT_TAB_OFFSET={build.flash_offset}                                                                                                        
 
+################################################################################
+# Generic L0
+GenL0.name=Generic STM32L0 series
+
+GenL0.build.vid=0x0483
+GenL0.build.core=arduino
+GenL0.build.board=GenL0
+GenL0.build.extra_flags={build.usb_flags} -DUSBD_USE_CDC -D{build.product_line} {build.enable_usb} {build.xSerial} -D__CORTEX_SC=0
```
[the remark this answers is not in the document]
Comment (fpistm) [reply boards.txt]: Still 
```suggestion
GenL0.build.extra_flags=-D{build.product_line} {build.enable_usb} {build.xSerial} -D__CORTEX_SC=0
```

File: variants/THUNDER_PACK/PinNamesVar.h [hunk 15]
```diff
@@ -0,0 +1,31 @@
+/* SYS_WKUP */
+#ifdef PWR_WAKEUP_PIN1
+SYS_WKUP1 = PA_0,
+#endif
+#ifdef PWR_WAKEUP_PIN2
+SYS_WKUP2 = PC_13,
```
[the remark this answers is not in the document]
Comment (fpistm) [reply variants/THUNDER_PACK/PinNamesVar.h]: To fix build issue with astyle kept 
`SYS_WKUP2 = PC_13`

File: variants/THUNDER_PACK/PeripheralPins.c [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+ * Copyright (c) 2019, STMicroelectronics
+ * All rights reserved.
+ *
+ * This software component is licensed by ST under BSD 3-Clause license,
+ * the "License"; You may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at:
+ *                        opensource.org/licenses/BSD-3-Clause
+ *
+ *******************************************************************************
+ * Automatically generated from STM32L072C(B-Z)Tx.xml
```
[the remark this answers is not in the document]
Comment (fpistm) [reply variants/THUNDER_PACK/PeripheralPins.c]: I guess you remove lines with pins not connected ?

[the remark this answers is not in the document]
Comment (jgillick) [reply variants/THUNDER_PACK/PeripheralPins.c]: Yup. Should I have commented those out instead?

[the remark this answers is not in the document]
Comment (fpistm) [reply variants/THUNDER_PACK/PeripheralPins.c]: No, if the pins are not available, that's fine.
